File: Bulwark.Core/Engine/AttackAnnotator.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;
using Bulwark.Core.Models;

namespace Bulwark.Core.Engine;

/// <summary>
/// ATT&CK 技战术标注器:从证据链各条目的描述里提取 MITRE ATT&CK 技战术编号
/// (形如 T1059 / T1218.010),用 <see cref="AttackCatalog"/> 解析为「编号 + 名称」,
/// 写回到每条证据的 <see cref="Evidence.Technique"/>,并在事件上汇总去重的
/// <see cref="SecurityEvent.Techniques"/>。
///
/// 采用"从原因文本提取"的方式:多个分析器(LolbinAnalyzer 等)已在原因里标注了 T 编号,
/// 这样无需逐个改造分析器即可统一产出结构化技战术标签。零额外检测逻辑,纯标注。
/// </summary>
public static class AttackAnnotator
{
    private static readonly Regex TechniqueRegex =
        new(@"T\d{4}(?:\.\d{3})?", RegexOptions.Compiled | RegexOptions.IgnoreCase);

    /// <summary>扫描事件证据链,填充每条证据的技战术字段并汇总到 <see cref="SecurityEvent.Techniques"/>。</summary>
    public static void Annotate(SecurityEvent e)
    {
        if (e.EvidenceChain is not { Count: > 0 }) return;

        // 保持首次出现顺序的去重集合
        var seen = new HashSet<string>(System.StringComparer.OrdinalIgnoreCase);
        var techniques = new List<string>();

        foreach (var ev in e.EvidenceChain)
        {
            if (string.IsNullOrEmpty(ev.Description)) continue;

            var m = TechniqueRegex.Match(ev.Description);
            if (!m.Success) continue;

            string id = m.Value.ToUpperInvariant();
            var t = AttackCatalog.Lookup(id);

            // 把(子编号优先的)规范编号与名称写回证据
            string canonicalId = t?.Id ?? id;
            ev.Technique ??= canonicalId;
            ev.TechniqueName ??= t?.Name;

            // 用原始命中的编号做去重键,保留子技战术粒度
            if (seen.Add(id))
                techniques.Add(AttackCatalog.Describe(id));
        }

        if (techniques.Count > 0)
            e.Techniques = techniques;
    }
}

File: cpp/shared/src/engine/AttackAnnotator.cpp
#include "bulwark/engine/AttackAnnotator.h"
#include "bulwark/engine/AttackCatalog.h"
#include <QSet>
#include <QStringList>
#include <QRegularExpression>

namespace bulwark::engine {

void AttackAnnotator::annotate(bulwark::SecurityEvent& e) {
    if (e.evidenceChain.isEmpty()) return;

    static const QRegularExpression techniqueRe(
        QStringLiteral("T\\d{4}(?:\\.\\d{3})?"), QRegularExpression::CaseInsensitiveOption);

    QSet<QString> seen;      // 去重键(大写编号)
    QStringList techniques;  // 保持首次出现顺序

    for (bulwark::Evidence& ev : e.evidenceChain) {
        if (ev.description.isEmpty()) continue;

        const auto m = techniqueRe.match(ev.description);
        if (!m.hasMatch()) continue;

        const QString id = m.captured(0).toUpper();
        const auto t = AttackCatalog::lookup(id);

        // 规范编号(子编号优先)与名称写回证据(仅在未填充时)。
        const QString canonicalId = t.has_value() ? t->id : id;
        if (ev.technique.isEmpty()) ev.technique = canonicalId;
        if (ev.techniqueName.isEmpty() && t.has_value()) ev.techniqueName = t->name;

        if (!seen.contains(id)) {
            seen.insert(id);
            techniques << AttackCatalog::describe(id);
        }
    }

    if (!techniques.isEmpty())
        e.techniques = techniques;
}

} // namespace bulwark::engine
